src/learning/nnet/net_layer.rs

```diff
@@ -0,0 +1,159 @@
+
+use linalg::{Matrix, MatrixSlice, BaseMatrix, BaseMatrixMut};
+use rulinalg::utils;
+
+use learning::LearningResult;
+use learning::error::{Error, ErrorKind};
+use learning::toolkit::activ_fn::ActivationFunc;
+
+use rand::thread_rng;
+use rand::distributions::Sample;
+use rand::distributions::normal::Normal;
+
+use std::fmt::Debug;
+
+/// Trait for neural net layers
+pub trait NetLayer : Debug {
+    /// The result of propogating data forward through this layer
+    fn forward(&self, input: &Matrix<f64>, params: MatrixSlice<f64>) -> LearningResult<Matrix<f64>>;
+
+    /// The gradient of the output of this layer with respect to its input
+    fn back_input(&self, out_grad: &Matrix<f64>, input: &Matrix<f64>, params: MatrixSlice<f64>) -> Matrix<f64>;
+    
+    /// The gradient of the output of this layer with respect to its parameters
+    fn back_params(&self, out_grad: &Matrix<f64>, input: &Matrix<f64>, params: MatrixSlice<f64>) -> Matrix<f64>;
+
+    /// The default value of the parameters of this layer before training
+    fn default_params(&self) -> Vec<f64>;
+
+    /// The shape of the parameters used by this layer
+    fn param_shape(&self) -> (usize, usize);
+
+    /// The number of parameters used by this layer
+    fn num_params(&self) -> usize {
+        let shape = self.param_shape();
+        shape.0 * shape.1
+    }
+}
+
+/// Linear network layer
+///
+/// Represents a fully connected layer with optional bias term
+///
+/// The parameters are a matrix of weights of size I x O
+/// where O is the dimensionality of the output and I the dimensionality of the input
+#[derive(Debug, Clone, Copy)]
+pub struct Linear { 
+    /// The number of dimensions of the input
+    input_size: usize,
+    /// The number of dimensions of the output
+    output_size: usize,
+    /// Whether or not to include a bias term
+    has_bias: bool,
+}
+
+impl Linear {
+    /// Construct a new Linear layer
+    pub fn new(input_size: usize, output_size: usize) -> Linear {
+        Linear {
+            input_size: input_size + 1, 
+            output_size: output_size,
+            has_bias: true
+        }
+    }
+
+    /// Construct a Linear layer with a bias term
+    pub fn without_bias(input_size: usize, output_size: usize) -> Linear {
+        Linear {
+            input_size: input_size, 
+            output_size: output_size,
+            has_bias: false
+        }
+    }
+}
+
+impl NetLayer for Linear {
+    /// Computes a matrix product
+    ///
+    /// input should have dimensions N x I
+    /// where N is the number of samples and I is the dimensionality of the input
+    fn forward(&self, input: &Matrix<f64>, params: MatrixSlice<f64>) -> LearningResult<Matrix<f64>> {
+        if self.has_bias {
+            if input.cols()+1 != params.rows() {
+                Err(Error::new(ErrorKind::InvalidData, "The input had the wrong number of columns"))
+            } else {
+                Ok(&input.hcat(&Matrix::ones(input.rows(), 1)) * &params)
```

Comment:
I'm not an expert in machine learning so the question might be stupid. 
Why are the bias appended at the end of inputs instead of the beginning? Is it a mathematical trick or just a different convention?